### ISSUE TITLE
Data Management API

### ISSUE THREAD

**Issue by DominikPinsel** [eclipse-dataspaceconnector/DataSpaceConnector#476]
Create a API extension that can be used to manage data of the connector.
For starters make it possible to create/update/delete `Assets`, `DataAddress` and `ContractDefinitions`, as these classes are mandatory to execute the offer-, negotiation- and transfer-flow.  
This extension would make it possible to create a data loader launcher, whose API then may be used by a CLI or any other application.

The most important missing piece, is probably the possibility to update and delete data. I'd suggest do rename the `DataLoder` to `Repository` and add the functionality there. What do you think? 

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

**Comment by jimmarino** [comment]
As I mentioned in the other discussion threads, these are all distinct issues from a semantic and security standpoint and I don't think they should be mixed:

1. Loading data into the system
2. Creating and editing ContractDefinitions
3. Removing Assets/DataAdresses

We should discuss these issues first (particularly 2-3 as 1 is mostly complete except for updates), and once we have a plan, file issues.

I don't think we should rename DataLoader to Repository. We designed it as a data loader to support bulk operations and to separate the semantic of loading data into the system from simple CRUD operations. Given the complexities around removing Assets, a generic CRUD repository won't fit the bill and we will need to account for this case.

**Comment by DominikPinsel** [comment]
> As I mentioned in the other discussion threads, these are all distinct issues from a semantic and security standpoint and I don't think they should be mixed:
> 
> 1. Loading data into the system
> 2. Creating and editing ContractDefinitions
> 3. Removing Assets/DataAdresses
> 
> We should discuss these issues first (particularly 2-3 as 1 is mostly complete except for updates), and once we have a plan, file issues.
> 
> I don't think we should rename DataLoader to Repository. We designed it as a data loader to support bulk operations and to separate the semantic of loading data into the system from simple CRUD operations. Given the complexities around removing Assets, a generic CRUD repository won't fit the bill and we will need to account for this case.

Do you have a counter proposal? I assume with the data plane the user will want to 
- create, update and delete contract definitions
- create, update and delete assets

**Comment by jimmarino** [comment]
The data plane is a completely separate issue and doesn't relate to this set of issues. FWIW I have the core data plane framework (DPF) implemented. It needs to be rebased against main and then it should be good to start experimenting with. 

Creating contract definitions, assets, and data addresses are already implemented by the DataLoader. 

We need to provide the ability to perform updates, deletes, and possibly versioning and decommissioning of various data types in the system. As I said, these need to be discussed individually and some will require the implementation of very specific semantics, e.g. Assets.

As a counterproposal, I recommend:

1. Address #439 for Milestone 2 
2. Implement something similar to CosmosApi for SQL backends for Milestone 2
3. Have a roadmap and milestone discussion for when update, delete, and retention of various data types can potentially be scheduled. 
4. On the base of 1-3, create a design for updating and deleting ContractDefinitions. The loader should be able to handle updates. I chose this type first as it will be the easiest to deal with. We may even be able to sneak this into Milestone 2, although we would need to assess the impact on the overall schedule.
5. Identify use cases for Asset management and retention. Once we fully understand that, create a design to implement support for those use cases. 
6. Once 4 & 5 are more thoroughly understood, schedule those features for a milestone deliverable and work on them.

As I said, a generic CRUD repository is not going to be sufficient for the management assets. Jumping into coding before Steps 1-6 are undertaken is putting the cart before the horse.

**Comment by DominikPinsel** [comment]
> The data plane is a completely separate issue and doesn't relate to this set of issues. FWIW I have the core data plane framework (DPF) implemented. It needs to be rebased against main and then it should be good to start experimenting with.
>
> Creating contract definitions, assets, and data addresses are already implemented by the DataLoader.
> We need to provide the ability to perform updates, deletes, and possibly versioning and decommissioning of various data types in the system. As I said, these need to be discussed individually and some will require the implementation of very specific semantics, e.g. Assets.
>
> As a counterproposal, I recommend:
>
> 1. Address Provide library to support postgresql as data retention site (closes #438) #439 for Milestone 2

At the time of writing there are five pull request open for the integration of the PostgreSQL. I could open more, but as they depend on each other I would rather not. But I will open additional ones, once we start closing the existing PRs.

> 2. Implement something similar to CosmosApi for SQL backends for Milestone 2

Isn't a library for PostgreSQL already done in PR 438? Maybe I don't understand. Please point out the differences
between the setwo libraries.

> 3. Have a roadmap and milestone discussion for when update, delete, and retention of various data types can potentially be scheduled.

Do you have an example roadmap, where this was done for another issue? I would then do something similar.

> 4. On the base of 1-3, create a design for updating and deleting ContractDefinitions. The loader should be able to handle updates. I chose this type first as it will be the easiest to deal with. We may even be able to sneak this into Milestone 2, although we would need to assess the impact on the overall schedule.

You suggested that the `DataLoader` should not only do create, but also update operations. Does this mean you agree to my proposal, to refactor the loader and add all basic CRUD operations to it?

> 5. Identify use cases for Asset management and retention. Once we fully understand that, create a design to implement support for those use cases.

I don't understand how a use case may affect general mutating of resources. Do you have an example of a use case, that restricts in some way a basic creation, update or delete operation? Whether its deletion or soft deletion should not affect the interface.

> 6. Once 4 & 5 are more thoroughly understood, schedule those features for a milestone deliverable and work on them.
> 7. As I said, a generic CRUD repository is not going to be sufficient for the management of assets. Jumping into coding before Steps 1-6 are undertaken is putting the cart before the horse.

I agree that a CRUD repository is not sufficient for the management of assets. But that's not in the scope of the CRUD repositories, the management of assets should be done by the service layer above.

**Comment by jimmarino** [comment]
I'm afraid we are simply going in circles. Let me try and clarify.

>> As a counterproposal, I recommend:

>> Address Provide library to support postgresql as data retention site (closes Lib - Postgresql - Provide library to support postgresql as data retention site #438) Provide library to support postgresql as data retention site (closes #438) #439 for Milestone 2

> At the time of writing there are five pull request open for the integration of the PostgreSQL. I could open more, but as they depend on each other I would rather not. But I will open additional ones, once we start closing the existing PRs.

My point is exactly the opposite: please engage with @paullatzelsperger and address his proposed changes to #438 before opening up more PRs that rely on the former. 

>> Implement something similar to CosmosApi for SQL backends for Milestone 2
> Isn't a library for PostgreSQL already done in PR 438? Maybe I don't understand. Please point out the differences
between the setwo libraries.

This has already been pointed out in Paul's proposal and his explanation in the related discussion. Please refer to that.

>> Have a roadmap and milestone discussion for when update, delete, and retention of various data types can potentially be scheduled.
> Do you have an example roadmap, where this was done for another issue? I would then do something similar.
 
Sure, the roadmap process we agreed to and followed for Milestone 1. We have already discussed on many occasions that the project will be following that process for future releases where we organize features into milestone releases and manage our development efforts accordingly. 

We have already discussed Milestone 2 issues following this process. For example, the Data Plane discussions, which you were part of. Markus will be circulating a document soon that outlines the specific roadmap for people to provide feedback. Note that #438 is on the roadmap for Milestone 2 but Asset retention and management is not. Given all of the other work we have on our plate for Milestone 2, I think this will need to be in a future release.

>You suggested that the DataLoader should not only do create, but also update operations. Does this mean you agree to my proposal, to refactor the loader and add all basic CRUD operations to it?

Please re-read my statements about the design of the DataLoader. Bulk load/update is different from reading and deleting. To be clear, I do not agree with the proposal to essentially re-design and re-write the DataLoader as a CRUD repository.

>> Identify use cases for Asset management and retention. Once we fully understand that, create a design to implement support for those use cases.
>I don't understand how a use case may affect general mutating of resources. Do you have an example of a use case, that restricts in some way a basic creation, update or delete operation? Whether its deletion or soft deletion should not affect the interface.

By "use case for Asset management and retention", I mean the scenarios and expectations concerning when an Asset is updated or "removed" from the system. For example, what happens to contract agreements? How is this handled, etc. This has nothing to do with the technical definition of an interface or other code to remove an asset entry from a storage system. Again, that is putting the cart before the horse. We need to understand the requirements before we jump into coding. 

>I agree that a CRUD repository is not sufficient for the management of assets. But that's not in the scope of the CRUD repositories, the management of assets should be done by the service layer above.

That is different than what you posted in this issue at the outset:

>The most important missing piece, is probably the possibility to update and delete data. I'd suggest do rename the DataLoder to Repository and add the functionality there. What do you think?

Now I am completely confused about what the issue is that needs to be addressed and what the exact proposal is.

My suggestion is to go back and follow the steps I outlined previously and start with a roadmap and requirements discussion.

**Comment by DominikPinsel** [comment]
> I'm afraid we are simply going in circles. Let me try and clarify.
>
> > > As a counterproposal, I recommend:
>
> > > Address Provide library to support postgresql as data retention site (closes Lib - Postgresql - Provide library to support postgresql as data retention site #438) Provide library to support postgresql as data retention site (closes #438) #439 for Milestone 2
>
> > At the time of writing there are five pull request open for the integration of the PostgreSQL. I could open more, but as they depend on each other I would rather not. But I will open additional ones, once we start closing the existing PRs.
>
> My point is exactly the opposite: please engage with @paullatzelsperger and address his proposed changes to #438 before opening up more PRs that rely on the former.

Done

> > > Implement something similar to CosmosApi for SQL backends for Milestone 2
> > > Isn't a library for PostgreSQL already done in PR 438? Maybe I don't understand. Please point out the differences
> > > between the setwo libraries.
>
> This has already been pointed out in Paul's proposal and his explanation in the related discussion. Please refer to that.

I just did answer Paul's proposal. While this is definitely a possible feature for the future, I don't see supporting
all kinds of SQL backends as part of the existing PR nor in the scope of this issue.

> > > Have a roadmap and milestone discussion for when update, delete, and retention of various data types can potentially be scheduled.
> > > Do you have an example roadmap, where this was done for another issue? I would then do something similar.
>
> Sure, the roadmap process we agreed to and followed for Milestone 1. We have already discussed on many occasions that the project will be following that process for future releases where we organize features into milestone releases and manage our development efforts accordingly.

Maybe then I just don't know of it. If you tell me where the roadmap for Milestone 2 is I'll make a proposal to update the missing pieces accordingly.

> We have already discussed Milestone 2 issues following this process. For example, the Data Plane discussions, which you were part of. Markus will be circulating a document soon that outlines the specific roadmap for people to provide feedback. Note that #438 is on the roadmap for Milestone 2 but Asset retention and management is not. Given all of the other work we have on our plate for Milestone 2, I think this will need to be in a future release.
>
> > You suggested that the DataLoader should not only do create, but also update operations. Does this mean you agree to my proposal, to refactor the loader and add all basic CRUD operations to it?
>
> Please re-read my statements about the design of the DataLoader. Bulk load/update is different from reading and deleting. To be clear, I do not agree with the proposal to essentially re-design and re-write the DataLoader as a CRUD repository.

Alright. I assume in this case the `Repository` interface would be something separate, used for data management. While the `DataLoader` interface stays as it is for bulk operations.

> > > Identify use cases for Asset management and retention. Once we fully understand that, create a design to implement support for those use cases.
> > > I don't understand how a use case may affect general mutating of resources. Do you have an example of a use case, that restricts in some way a basic creation, update or delete operation? Whether its deletion or soft deletion should not affect the interface.
>
> By "use case for Asset management and retention", I mean the scenarios and expectations concerning when an Asset is updated or "removed" from the system. For example, what happens to contract agreements? How is this handled, etc. This has nothing to do with the technical definition of an interface or other code to remove an asset entry from a storage system. Again, that is putting the cart before the horse. We need to understand the requirements before we jump into coding.

I have taken these things already in account and will write a new comment with further explanation.

> > I agree that a CRUD repository is not sufficient for the management of assets. But that's not in the scope of the CRUD repositories, the management of assets should be done by the service layer above.
>
> That is different than what you posted in this issue at the outset:
>
> > The most important missing piece, is probably the possibility to update and delete data. I'd suggest do rename the DataLoder to Repository and add the functionality there. What do you think?
>
> Now I am completely confused about what the issue is that needs to be addressed and what the exact proposal is.

I will also try to go here more in detail in a follow-up comment.

> My suggestion is to go back and follow the steps I outlined previously and start with a roadmap and requirements discussion.

**Comment by DominikPinsel** [comment]
# Data Management API

The Data Management API allows the management of the EDC domain objects. For starters this domain objects should only include the `Asset` and the `ContractDefinition` as these are the most important objects to get the offer-, negotiate-
and transfer-flow running.

## Components

### Data Management Controller

The controller supports the following HTTP methods

- GET Query Asset(s)
- UPDATE Asset
- DELETE Asset
- CREATE Asset
- GET Find all ContractDefinition
- CREATE ContractDefinition
- UPDATE ContractDefinition
- DELETE ContractDefinition

Additionally, the controller will need to support a basic security mechanism like the control API.

#### Method Details

##### QUERY Asset(s)

Assets may be queried using the QueryParameters of the URL. Example Query:

```
https://example.de/api/data/asset?asset:prop:id=1
```

Restrictions:

- Left and right operand can only be of type _String_.
- Operator is always _equals_.

##### GET Find all ContractDefinition

The `ContractDefinitionStore` has not solved the problem with querying `ContractDefinition`. Therefore, this feature
will also be out of scope of this issue.

### Data Management Service

The `DataManagementService` will be part of a new core module. This is necessary, to run the Data Management as
standalone application (for security reasons).

The `DataManagementService`

- will be used by the DataManagement Controller and
- functions as Facade for the different `Repository` implementations
- add operation validation (not now but maybe in the future)

#### Operation Validation

Or why it's not necessary for now.

Let's have a look at potential side effects for the offering-, negotiation- and transfer-flow.

###### Contract Offering Flow

If an `Asset` or `ContractDefinition` is changed the corresponding offers will look different. As (currently) all offers
are generated on the fly and not persisted in some kind of central entity (like an IDS broker), in case of an `Asset`
or `ContractDefinition` change there will be simply updated offers generated in the future without any side effects to
expect.

###### Contract Negotiation Flow

If an `Asset` is changed during a contract negotiation for a corresponding offer, it should be rejected as the proposed
contract offer no longer matches the one, that would be generated. This behavior should already be implemented
accordingly.

If an `ContractDefinition` is changed during a contract negotiation for a corresponding offer, it should be rejected as
the proposed offer no longer matches the one, that would be generated. This behavior should already be implemented
accordingly.

###### Data Transfer Flow

If an `Asset` or `ContractDefinition` is changed during a data transfer flow it won't affect the data transfer. As the
data transfer will use the `ContractAgreement` to validate the transfer, which contains all the necessary contract
information and `Assets`, of the time the `ContractAgreement` was created.

### Repository Interface

The `Repository` interface will be implemented for different kind of storage systems and is used by
the `DataManagementService`. It will support the following operations:

- create, update delete`Assets` and `ContractDefinitions`
- query `Assets` using `Criteria`
- get all `ContractDefintions`

## Necessary Refactoring

### Move DataAddress in the Asset object

At the time of writing the `DataAddress` has no link to the `Asset` objects. This is problematic when the user would
want to access the `DataAddress` using the Data Management API. Because, after getting the address, the user would not be able to tell to which `Asset` it belongs to.

There are different solutions possible. One would be to add the identifier of one object to the other one, so that they
are loosely coupled. But that just adds unnecessary complexity, it is probably easier to move the `DataAddress` into
the `Asset`.

# Out of scope

The following topics are necessary to solve, but not in the scope of this issue, as it's already pretty big.

## Deletion Strategy

There are probably different kinds of scenarios where

- a soft delete would be the preferred deletion method.
- deleted resources should be archived.
- soft deleted resources should be deleted after some kind of trigger.
- deletion strategy should be configurable.

The overall strategy for the different resources must be defined at a later time. For this issue a hard delete is fine.

## Querying of Contract Definitions

The querying of contract definitions comes with a whole set of new problems and will therefore not be solved within this
issue.

## Versioning

Versioning of domain objects and how to handle these different versions will not be solved within this issue.

**Comment by MoritzKeppler** [comment]
Some thoughts:

We should draw a diagram how a DataLoader plays together with the other EDC processes. And especially document the reasoning for having it in a separate process. Maybe we have that and I just lost the overview. If not: I volunteer to work on that maybe together with @ronjaquensel 

People are asking us for a REST API that can be called from a javascript UI to manage assets and contracts. Wouldn't that result in a DataLoader Runtime that offers such an API (in parallel to the CLI runtime)? 

What would be a good setup for the roadmap and milestone discussion you mentioned, @jimmarino?

btw. you also mentioned the DPF - is that already in the code / visible as (draft) PR? Sounds great....

<sub> Moritz Keppler <moritz.keppler@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sub>

**Comment by DominikPinsel** [comment]
Another issue with the `DataAddress`: It needs to be resolvable for `Asses`, that are part of an `ContractAgreement`.

**Comment by DominikPinsel** [comment]
Closing the orignal issue as duplicated. It contains a small subset of the Data-Mgmt-Features described in the issues numbering 583 to 593 from @paullatzelsperger.
